modules/mvt/src/lib/binary-vector-tile/vector-tile-layer.js

```diff
@@ -0,0 +1,79 @@
+// This code is forked from https://github.com/mapbox/vector-tile-js under BSD 3-clause license.
+
+import VectorTileFeature from './vector-tile-feature.js';
+
+export default class VectorTileLayer {
+  constructor(pbf, end) {
+    // Public
+    this.version = 1;
+    this.name = null;
+    this.extent = 4096;
+    this.length = 0;
+
+    // Private
+    this._pbf = pbf;
+    this._keys = [];
+    this._values = [];
+    this._features = [];
+
+    pbf.readFields(readLayer, this, end);
+
+    this.length = this._features.length;
+  }
+
+  // return feature `i` from this layer as a `VectorTileFeature`
+  feature(i, firstPassData) {
+    if (i < 0 || i >= this._features.length) {
+      throw new Error('feature index out of bounds');
+    }
+
+    this._pbf.pos = this._features[i];
+
+    const end = this._pbf.readVarint() + this._pbf.pos;
+    return new VectorTileFeature(
+      this._pbf,
+      end,
+      this.extent,
+      this._keys,
+      this._values,
+      firstPassData
+    );
+  }
+}
+
+function readLayer(tag, layer, pbf) {
+  if (tag === 15) layer.version = pbf.readVarint();
+  else if (tag === 1) layer.name = pbf.readString();
+  else if (tag === 5) layer.extent = pbf.readVarint();
+  else if (tag === 2) layer._features.push(pbf.pos);
+  else if (tag === 3) layer._keys.push(pbf.readString());
+  else if (tag === 4) layer._values.push(readValueMessage(pbf));
+}
+
+function readValueMessage(pbf) {
+  let value = null;
+  const end = pbf.readVarint() + pbf.pos;
+
+  while (pbf.pos < end) {
+    const tag = pbf.readVarint() >> 3;
+
+    value =
+      tag === 1
+        ? pbf.readString()
+        : tag === 2
+          ? pbf.readFloat()
+          : tag === 3
+            ? pbf.readDouble()
+            : tag === 4
+              ? pbf.readVarint64()
+              : tag === 5
+                ? pbf.readVarint()
+                : tag === 6
+                  ? pbf.readSVarint()
+                  : tag === 7
+                    ? pbf.readBoolean()
+                    : null;
```

Comment:
In my opinion the successive indentation makes things less clear, and I think something like below is much more readable. Also, it seems a little surprising to me that each iteration of the `while` loop overwrites `value` instead of appending to it. Maybe worth a comment?

```js
let value = null;
switch (tag) {
  case 1:
    value = pbf.readString();
    break;
  case 2:
    value = pbf.readFloat();
    break;
  case 3:
    value = pbf.readDouble();
    break;
  case 4:
    value = pbf.readVarint64();
    break;
  case 5:
    value = pbf.readVarint();
    break;
  case 6:
    value = pbf.readSVarint();
    break;
  case 7:
    value = pbf.readBoolean();
    break;
  default:
    value = null;
    break;
}
```

Reply:
Yes, I raised an eyebrow when I saw this too, but as per earlier comment, I would rather not change the original implementation if not necessary